sycl/include/sycl/ext/oneapi/experimental/graph.hpp

```diff
@@ -0,0 +1,222 @@
+
+  std::function<void(sycl::handler &)> my_body;
+
+  void exec(sycl::queue q) {
+    std::vector<sycl::event> __deps;
+    for (auto i : my_predecessors)
+      __deps.push_back(i->get_event());
+    my_event = q.submit(wrapper{my_body, __deps});
+  }
+
+  void register_successor(node_ptr n) {
+    my_successors.push_back(n);
+    n->register_predecessor(node_ptr(this));
+  }
+
+  void register_predecessor(node_ptr n) { my_predecessors.push_back(n); }
+
+  sycl::event get_event(void) { return my_event; }
+
+  template <typename T>
+  node_impl(graph_ptr g, T cgf)
+      : is_scheduled(false), my_graph(g), my_body(cgf) {}
+
+  // Recursively adding nodes to execution stack:
+  void topology_sort(std::list<node_ptr> &schedule) {
+    is_scheduled = true;
+    for (auto i : my_successors) {
+      if (!i->is_scheduled)
+        i->topology_sort(schedule);
+    }
+    schedule.push_front(node_ptr(this));
+  }
+};
+
+struct graph_impl {
+  std::set<node_ptr> my_roots;
+  std::list<node_ptr> my_schedule;
+  // TODO: Change one time initialization to per executable object
+  bool first;
+
+  graph_ptr parent;
+
+  void exec(sycl::queue q) {
+    if (my_schedule.empty()) {
+      for (auto n : my_roots) {
+        n->topology_sort(my_schedule);
+      }
+    }
+    for (auto n : my_schedule)
+      n->exec(q);
+  }
+
+  void exec_and_wait(sycl::queue q) {
+    if(first) {
+        exec(q);
+        first=false;
+    }
+    q.wait();
+  }
+
+  void add_root(node_ptr n) {
+    my_roots.insert(n);
+    for (auto n : my_schedule)
+      n->is_scheduled = false;
+    my_schedule.clear();
+  }
+
+  void remove_root(node_ptr n) {
+    my_roots.erase(n);
+    for (auto n : my_schedule)
+      n->is_scheduled = false;
+    my_schedule.clear();
+  }
+
+  graph_impl() : first(true) {}
+};
+
+} // namespace detail
+
+struct node {
+  detail::node_ptr my_node;
+  detail::graph_ptr my_graph;
+
+  template <typename T>
+  node(detail::graph_ptr g, T cgf)
+      : my_graph(g), my_node(new detail::node_impl(g, cgf)){};
+  void register_successor(node n) { my_node->register_successor(n.my_node); }
+  void exec(sycl::queue q, sycl::event = sycl::event()) { my_node->exec(q); }
+
+  void set_root() { my_graph->add_root(my_node); }
+};
+
+enum class graph_state{
+  modifiable,
+  executable
+};
+
+template<graph_state State=graph_state::modifiable>
+class command_graph {
+public:
+  // Adding empty node with [0..n] predecessors:
+  node add(const std::vector<node> &dep = {});
+
+  // Adding device node:
+  template <typename T>
+  node add(T cgf, const std::vector<node> &dep = {});
+
+  // Adding dependency between two nodes.
+  void make_edge(node sender, node receiver);
+
+  // TODO: Extend queue to directly submit graph
+  void exec_and_wait(sycl::queue q);
+
+  command_graph<graph_state::executable> finalize(const sycl::context &syclContext) const;
+
+  command_graph() : my_graph(new detail::graph_impl()) {}
+
+private:
+  detail::graph_ptr my_graph;
```

Comment:
Convention seems to be PascalCase for member variables, is there a reason not to follow it here?
```suggestion
  detail::graph_ptr MGraph;
```

Reply:
done.